pkg/download/downloader.go

```diff
@@ -162,20 +163,44 @@ func GetInsecure(uri, dir string, fetcher Fetcher) error {
 	return extractArchive(name, dir, body, size)
 }
 
-func extractArchive(filename, dst string, r io.ReaderAt, size int64) error {
+func extractContentType(at io.ReaderAt) (string, error) {
+	buf := make([]byte, 512)
+	n, err := at.ReadAt(buf, 0)
+	if err != nil && err != io.EOF {
+		return "", errors.Wrap(err, "failed to read first 512 bytes")
+	}
+	if n < 512 {
+		glog.V(5).Infof("Did only read %d of 512 bytes to determine the file type", n)
+	}
+	return http.DetectContentType(buf[:n]), nil
+}
+
+type extractor func(targetDir string, read io.ReaderAt, size int64) error
+
+var defaultExtractors = map[string]extractor{
+	"application/zip":      extractZIP,
+	"application/tar+gzip": extractTARGZ,
+	"application/x-gzip":   extractTARGZ,
```

Comment:
Not all gzip uncompress into tarballs though.

GitHub returns `application/x-gzip` for tarballs, but https://superuser.com/questions/901962/what-is-the-correct-mime-type-for-a-tar-gz-file says `application/gzip` would be recommended out in the wild. 🤔 

I guess we can take a risk and assume all GZIPs given to krew will be tarballs and not .zip.gz or something. :)

Reply:
It seems like Go only detects `application/x-gzip` and `application/zip` https://golang.org/src/net/http/sniff.go#L147